### PR TITLE
fix: Consistent tooltips for menu entries and toolbar buttons

### LINKED ISSUE #1605: Tooltips in toolbar not helpful
For beginner contributors looking for a _Good First Issue_: See [comment in this discussion](https://github.com/bit-team/backintime/issues/1605#issuecomment-1893537459) for instructions how to solve the issue.

The tooltips in our mainwindow are connected to `QActions` instead of specific buttons and menu entries. These actions are then connected to buttons and menu entries. The consequence is that menu entries and buttons in the toolbar do share the same tooltip.

Peek 2024-01-08 15-05

The menu entry "Take a snapshot" describe itself by its label. The tooltip then is used to give background or more detailed information.
The toolbar button (usually) do not have a label string. Because of that most (?) users do expect that the tooltip of such a button do describe its function in a short way. The same way a menu entry label would do.

In short: The tooltip of the button should be "Take a snapshot" or something else.

There are some solutions we could think about. It is not urgent. I just wanted to document it.

### PR DESCRIPTION
**Pull request by JATIN-RAGHAV**
I made the suggested changes for both snapshot and restore button. If it is not correct then please point out the errors. I wil try to fix them too.

Fix #1605

### CODE REVIEW

**Comment by buhtz** [comment]
Dear JATIN-RAGHAV,
thanks for this contribution and thank you for sticking with us. No need to hurry. I could also have been a little more patient.

## Some general aspects about contribution

1. Contributors should read `CONTRIBUTING.md`. There are some steps described that make it easier for us maintainers work with Pull Requests from external contributors. Now I modified that file a bit in hope to make some aspects clearer (see next bullet point).

2. Next time when you create another PR please create a separate branch for each contribution. Don't work on `dev` or `main`. Before create a branch based on `JATIN-RAGHAV:dev`  make sure that your local copy is synchronized with our upstream `bit-team:dev`.

```
$ git checkout dev
$ git checkout -b mybranch
$ [do your coding]
$ git commit -am 'commit message'
$ git push
```

3. For this specific PR just leave it as it is. I will manage.

4. You could have kept your first PR open and modified it. You can work on PRs that are still open. Just commit and push to the related branch. The modification will be noticed in the PR automatically.

## Your Code & Next steps

Feel free to ask and let me know if I should help you further with some tasks.

1. You might have noticed that I added some extra commits to the branch at your fork. Because of that you have to "refresh" your local repository. Do `git fetch --pull --prune && git pull` to synchronize the local copy of your repository with your remote repository `JATIN-RAGHAV/backintime`.

2. Test your version of BIT not only by unittests but also by starting it and checking the produced tooltips. You should find a typo in your code when starting BIT. That is also the reason why the TravisCI check is failing.

3. I added some review comments to your code.

## Your "indirect" contribution

I just want to mention that you contributed "indirect" to the next release (PR #1624). Your code inspired me to look for specific error (`undefinied-variable`) in other parts of the code. So I improved our linter (e.g. see `qt/tests/test_lint.py`) and discovered a bug in `qt/app.py`. Just because of your contribution! 😃 That is team work. You see there is much more value in your own contribution as you might notice.

## My learning

[Disabling pylint warnings in a specific scope](https://pylint.pycqa.org/en/latest/user_guide/messages/message_control.html) don't work for a whole file. A file is not a scope for pylint.

**Comment by buhtz** [comment]
Set this to DRAFT. After some research and to my understanding merging this (even if squashed) would destroy or corrupt the commit history somehow. I will think of a solution how to solve this without throwing this commits into the trashcan.

EDIT: Sorry for so much messy commits. I try to investigate a problem with pylint which throws "invalid-names" errors (for good reasons) on TravisCI but not on my local machine.

**Comment by buhtz** [comment]
@JATIN-RAGHAV: ~~Please pause working on this PR. I will first have to find a solution for the branching problem. I am also learning here. 😆~~

~~Please show me the output of the command `git remote -v` on your site.~~

**EDIT**:
I think it is messed up and you have to re create that PR again. Maybe a git-magician could solve the problem. Based on some experiments of my own please do the following.

    $ cd [into your local repo]
    $ git remote add upstream git@github.com:bit-team/backintime.git
    $ git remote -v
    origin  git@github.com:JATIN-RAGHAV/backintime.git (fetch)
    origin  git@github.com:JATIN-RAGHAV/backintime.git (push)
    upstream        git@github.com:bit-team/backintime.git (fetch)
    upstream        git@github.com:bit-team/backintime.git (push)

The output of the last command should look exactly like this. If not, please stop. Now do this:

    $ git checkout dev
    $ git checkout -b dirty
    $ git checkout dev
    $ git fetch --prune --all
    $ git pull upstream
    $ git checkout dirty
    $ git merge dev
    $ git push origin

Now look at your account at Microsoft GitHub. You should see a new branch named `dirty` in your own forked backintime repository.

Try to create a new PR based on that `JATIN-RAGHAV/dirty`.

**Comment by buhtz** [comment]
Dear JATIN-RAGHAV,
I am also still a learner and learned that it might not be a problem contributing using the "dev" branch. So just forget all my blabla about the git magic and stick to the main problem.

## Your Code & Next steps

Feel free to ask and let me know if I should help you further with some tasks.

1. You might have noticed that I added some extra commits to the branch at your fork. Because of that you have to "refresh" your local repository. Do `git fetch --pull --prune && git pull` to synchronize the local copy of your repository with your remote repository `JATIN-RAGHAV/backintime`.

2. Test your version of BIT not only by unittests but also by starting it and checking the produced tooltips. You should find a typo in your code when starting BIT. That is also the reason why the TravisCI check is failing.

5. I added some review comments to your code.

**Comment by buhtz** [comment]
I will take over the PR in some days if there is no response from @JATIN-RAGHAV anymore.

**Comment by buhtz** [comment]
As some people told me and to my own understanding it shouldn't be a problem to squash-merge dev into dev. But I am a bit irrational about that. So I would prefere if someone else would push the "Squash and merge" button. :rofl: :fearful: 

- Tooltip of toolbar buttons now is build from the underlying QActions "text" (menu label) and its "toolTip".
- Existing source strings are not modified. So now new translations needed.
- Only main toolbar is affected. The fileview toolbar is fine as it is but can be adapted if needed.
- The current solution take Right-to-Left (aka BIDI) languages or the reading direction into account.

Thanks to @JATIN-RAGHAV for contribution.

File: qt/app.py
Comment (aryoda) [inline qt/app.py:724]: I would have been much easier to "inject" the tooltip if Qt5 would have added the tooltip text to the `QAction` constructor. Now you have to loop over each action "just" to set the tooltip ;-)

File: qt/app.py
Comment (aryoda) [inline qt/app.py]: I would prefer to use the enum constant `Qt.LayoutDirection.RightToLeft` instead of the magic number `1`...

**Review by aryoda — APPROVED** [review]
@buhtz I am looking forward to our refactoring of the GUI into View and ViewModel to enable GUI testing (your `QActions` are the perfect preparation for this)

**Comment by buhtz** [comment]
OK, who wants to push the button? 😨

File: qt/app.py
Comment (aryoda) [inline qt/app.py:717]: I just realized that the enum also provides the constant `Qt.LayoutDirection.LayoutDirectionAuto` which would lead to applying LTR (which may be wrong). I have no idea how to cope with that constant...

**Comment by buhtz** [comment]
I shouldn't have skipped the last commit. Learning!!!

I investigate ...

**Comment by aryoda** [comment]
I think this has nothing to do with using an enum constant or the underlying value directly but the possible return values of `toolbar.layoutDirection()`

**Comment by buhtz** [comment]
I fixed the problem and comited it into the qt6migration branch.
https://github.com/bit-team/backintime/pull/1632

The problem was that pylint do not import `PyQt6.QtCore.Qt` and because of that is not able to know that `LayoutDirection` exist as a member in it.

Background: It is an outstanding advantage of pylint that it tries to import packages and determining its members. I am not aware of another linter trying this. That is why pylint is so slow. Because of security reasons pylint do not import C-extension packages. Qt is written in C. That is why we need to use `--extension-pkg-whitelist` to make pylint import such package explicit.
Another similar pitfall is that some packages to have dynamically generated members that do not exist on import time. For such members the option `--generated-members` do exist.

**Comment by aryoda** [comment]
Perhaps it is more secure to let Qt evaluate the layout direction via [`QGuiApplication::isLeftToRight()`](https://doc.qt.io/qt-6.2/qguiapplication.html#isLeftToRight)

**Comment by buhtz** [comment]
> Perhaps it is more secure to let Qt evaluate the layout direction via [`QGuiApplication::isLeftToRight()`](https://doc.qt.io/qt-6.2/qguiapplication.html#isLeftToRight)

Good idea. 865b3a3

**Pull request merged**
